packages/xpath/src/functions/xforms/node-set.ts

```diff
@@ -384,8 +386,39 @@ export const randomize = new NodeSetFunction(
 
 		const nodeResults = Array.from(results.values());
 		const nodes = nodeResults.map(({ value }) => value);
-		const seed = seedExpression?.evaluate(context).toNumber();
-
-		return seededRandomize(nodes, seed);
+		if (seedExpression === undefined) return seededRandomize(nodes);
+		const seed = seedExpression.evaluate(context);
+		const asNumber = seed.toNumber(); // TODO: There are some peculiarities to address: https://github.com/getodk/web-forms/issues/240
+		let finalSeed: number | bigint | undefined;
+		if (Number.isNaN(asNumber)) {
+			// Specific behaviors for when a seed value is not interpretable as numeric.
+			// We still want to derive a seed in those cases, see https://github.com/getodk/javarosa/issues/800
+			const seedString = seed.toString();
+			if (seedString === '') {
+				finalSeed = 0; // special case: JR behaviour
+			} else {
+				// any other string, we'll convert to a number via a digest function
+				finalSeed = toBigIntHash(seedString);
+			}
+		} else {
+			finalSeed = asNumber;
+		}
+		return seededRandomize(nodes, finalSeed);
 	}
 );
+
+function toBigIntHash(text: string): bigint {
+	// hash text with sha256, and interpret the first 64 bits of output
+	// (the first and second int32s ("words") of CryptoJS digest output)
+	// as a BigInt. Thus the entropy of the hash is reduced to 64 bits, which
+	// for some applications is sufficient.
+	// The underlying representations are big-endian regardless of the endianness
+	// of the machine this runs on, as is the equivalent JavaRosa implementation
+	// at https://github.com/getodk/javarosa/blob/ab0e8f4da6ad8180ac7ede5bc939f3f261c16edf/src/main/java/org/javarosa/xpath/expr/XPathFuncExpr.java#L718-L726
```

Comment:
```suggestion
/**
 * Hash text with sha256, and interpret the first 64 bits of output (the first
 * and second int32s ("words") of CryptoJS digest output) as a BigInt. Thus the
 * entropy of the hash is reduced to 64 bits, which for some applications is
 * sufficient. The underlying representations are big-endian regardless of the
 * endianness of the machine this runs on, as is the
 * {@link https://github.com/getodk/javarosa/blob/ab0e8f4da6ad8180ac7ede5bc939f3f261c16edf/src/main/java/org/javarosa/xpath/expr/XPathFuncExpr.java#L718-L726 | equivalent JavaRosa implementation}.
 */
const toBigIntHash = (text: string): bigint => {
```

As a JSDoc comment, this allows the same documentation to be accessed at the call site.

Switching to an arrow function is somewhat a nit, but it's generally preferable to avoid unnecessary `function` functions as they have confusing behavior. (Maybe that's also a thing we could lint?)

Reply:
I'll try to make the case for JSDoc. If you're open to reconsidering here, that would be excellent. If not, I think we should come back to this discussion as a team.

- - -

JSDoc comments are a standard designed to encode structured documentation about any symbol they're attached to.

The editor support alone goes well beyond linking to URLs. For example, the ability to reference documentation across modules are invaluable. Linking to _other symbols_ (both within and across modules) is also invaluable, both as a navigation tool and because **they can be kept up to date** as those symbols change.

Beyond editor support, being a standard for structured documentation and association with symbols, JSDoc can be used for _documentation output_. We are already using this to generate documentation for `@getodk/xforms-engine`. I'd quite like to expand that to other use cases.

I share your distaste for some of the syntax minutiae of JSDoc, and `@link` is particularly weird (I suspect this is because inline tags are relatively rare). But that distaste doesn't outweigh the overwhelming benefits of an extensible documentation standard which is widely adopted in tooling we already use. It is also widely adopted throughout this project, and across the ecosystem; which is to say, it is both locally and globally idiomatic.

I also noticed GitHub's odd presentation in a couple diff suggestions in this PR. It's worth noting that:

- That's not representative of how GitHub presents JSDoc in complete source
- It's not representative of how GitHub presents JSDoc in diffs broadly
- GitHub's syntax highlighting is notoriously inconsistent across various views
- The highlighting is applied to an incomplete (i.e. syntactically invalid) chunk of code, which likely exacerbates potential issues

Lastly, I am sensitive to poor authoring ergonomics. I'm a bit surprised to hear that your editor doesn't make adding/editing JSDoc comments _easier than single line comments_, as that's my experience in the editors I'm familiar with. If this is a major hangup, I'd be happy to help look into ways to make the authoring experience nicer for you.